drivers/ethernet/dsa_ksz8794.c

```diff
@@ -0,0 +1,579 @@
+/*
+ * Copyright (c) 2020 DENX Software Engineering GmbH
+ *               Lukasz Majewski <lukma@denx.de>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define LOG_MODULE_NAME dsa
+#define LOG_LEVEL CONFIG_ETHERNET_LOG_LEVEL
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#include <device.h>
+#include <kernel.h>
+#include <errno.h>
+#include <sys/util.h>
+#include <drivers/spi.h>
+#include <net/ethernet.h>
+#include <linker/sections.h>
+
+#include "dsa_ksz8794.h"
+
+static struct dsa_ksz8794_spi phy_spi;
+
+static void dsa_ksz8794_write_reg(struct dsa_ksz8794_spi *sdev,
+				  uint16_t reg_addr, uint8_t value)
+{
+	uint8_t buf[3];
+
+	const struct spi_buf tx_buf = {
+		.buf = buf,
+		.len = 3
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1
+	};
+
+	buf[0] = KSZ8794_SPI_CMD_WR | ((reg_addr >> 7) & 0x1F);
+	buf[1] = (reg_addr << 1) & 0xFE;
+	buf[2] = value;
+
+	spi_write(sdev->spi, &sdev->spi_cfg, &tx);
+}
+
+static void dsa_ksz8794_read_reg(struct dsa_ksz8794_spi *sdev,
+				 uint16_t reg_addr, uint8_t *value)
+{
+	uint8_t buf[3];
+
+	const struct spi_buf tx_buf = {
+		.buf = buf,
+		.len = 3
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1
+	};
+	struct spi_buf rx_buf = {
+		.buf = buf,
+		.len = 3
+	};
+
+	const struct spi_buf_set rx = {
+		.buffers = &rx_buf,
+		.count = 1
+	};
+
+	buf[0] = KSZ8794_SPI_CMD_RD | ((reg_addr >> 7) & 0x1F);
+	buf[1] = (reg_addr << 1) & 0xFE;
+	buf[2] = 0x0;
+
+	if (!spi_transceive(sdev->spi, &sdev->spi_cfg, &tx, &rx)) {
+		*value = buf[2];
+	} else {
+		LOG_DBG("Failure while reading register 0x%04x", reg_addr);
+		*value = 0U;
+	}
+}
+
+static bool dsa_ksz8794_port_link_status(struct dsa_ksz8794_spi *sdev,
+					 uint8_t port)
+{
+	uint8_t tmp;
+
+	if (port < KSZ8794_PORT1 || port > KSZ8794_PORT3)
+		return false;
+
+	dsa_ksz8794_read_reg(sdev, KSZ8794_STAT2_PORTn(port), &tmp);
+
+	return tmp & KSZ8794_STAT2_LINK_GOOD;
+}
+
+static bool dsa_ksz8794_link_status(struct dsa_ksz8794_spi *sdev)
+{
+	bool ret = false;
+	uint8_t i;
+
+	for (i = KSZ8794_PORT1; i <= KSZ8794_PORT3; i++) {
+		if (dsa_ksz8794_port_link_status(sdev, i)) {
+			LOG_INF("Port: %d link UP!", i);
+			ret |= true;
+		}
+	}
+
+	return ret;
+}
+
+static void dsa_ksz8794_soft_reset(struct dsa_ksz8794_spi *sdev)
+{
+	/* reset switch */
+	dsa_ksz8794_write_reg(sdev, KSZ8794_PD_MGMT_CTRL1,
+			      KSZ8794_PWR_MGNT_MODE_SOFT_DOWN);
+	k_busy_wait(1000);
+	dsa_ksz8794_write_reg(sdev, KSZ8794_PD_MGMT_CTRL1, 0);
+}
+
+static int dsa_ksz8794_spi_setup(struct dsa_ksz8794_spi *sdev, uint32_t freq,
+				 uint8_t cs_num)
+{
+	uint16_t timeout = 100;
+	uint8_t val[2], tmp;
+
+	/* SPI config */
+	sdev->spi_cfg.operation =
+#if DSA_DT_SPI_CPOL
```

Comment:
shouldn't you be able to get these from DTS?

Reply:
The DSA_DT_SPI_CPOL is already set according to DTS. Marcros for it can be found here:
`include/net/dsa.h`

Reply:
Please find updated version of this patch set - it now uses `DT_DRV_COMPAT`